src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs

```diff
@@ -0,0 +1,373 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry.Exporter.Geneva;
+
+internal class OtlpProtobufSerializer
+{
+    private const int LengthAndTagSize = 4;
+
+    private static readonly ConcurrentBag<List<Metric>> MetricListPool = new();
+
+    private readonly Dictionary<string, List<Metric>> scopeMetrics = new();
+
+    private int resourceMetricStartIndex;
+
+    private int previousScopeMetricsStartIndex;
+
+    private int previousMetricStartIndex;
+
+    private int previousIntrumentStartIndex;
+
+    private int previousMetricPointStartIndex;
+
+    internal IMetricDataTransport MetricDataTransport;
+
+    public OtlpProtobufSerializer(IMetricDataTransport metricDataTransport)
+    {
+        this.MetricDataTransport = metricDataTransport;
+    }
+
+    internal void SerializeAndSendMetrics(byte[] buffer, ref int currentPosition, Resource resource, in Batch<Metric> metricBatch)
+    {
+        foreach (var metric in metricBatch)
+        {
+            if (this.scopeMetrics.TryGetValue(metric.MeterName, out var metricList))
+            {
+                metricList.Add(metric);
+            }
+            else
+            {
+                if (MetricListPool.TryTake(out var newList))
+                {
+                    newList.Add(metric);
+                    this.scopeMetrics[metric.MeterName] = newList;
+                }
+                else
+                {
+                    newList = new List<Metric>();
+                    newList.Add(metric);
+                    this.scopeMetrics[metric.MeterName] = newList;
+                }
+            }
+        }
+
+        // Serialize
+        this.SerializeResourceMetrics(buffer, ref currentPosition, resource);
```

Comment:
Same thing as this comment: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1596/files#r1529583054